pkg/watchdog/watchdog.go

```diff
@@ -0,0 +1,253 @@
+/*
+ * This file is part of the kubevirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2017 Red Hat, Inc.
+ *
+ */
+
+package watchdog
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	k8sv1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	diskutils "kubevirt.io/kubevirt/pkg/ephemeral-disk-utils"
+	"kubevirt.io/kubevirt/pkg/logging"
+	"kubevirt.io/kubevirt/pkg/precond"
+	"kubevirt.io/kubevirt/pkg/virt-handler/virtwrap/api"
+)
+
+func NewWatchdogListWatchFromClient(virtShareDir string, watchdogTimeout int) cache.ListerWatcher {
+
+	d := &WatchdogListWatcher{
+		fileDir:                  WatchdogFileDirectory(virtShareDir),
+		backgroundWatcherStarted: false,
+		watchdogTimeout:          watchdogTimeout,
+	}
+	return d
+}
+
+func WatchdogFileDirectory(baseDir string) string {
+	return filepath.Clean(baseDir) + "/watchdog-files"
+}
+
+func WatchdogFileFromNamespaceName(baseDir string, namespace string, name string) string {
+	return filepath.Clean(baseDir) + "/watchdog-files/" + namespace + "_" + name
+}
+
+func WatchdogFileRemove(baseDir string, vm *v1.VirtualMachine) error {
+	namespace := precond.MustNotBeEmpty(vm.GetObjectMeta().GetNamespace())
+	domain := precond.MustNotBeEmpty(vm.GetObjectMeta().GetName())
+
+	file := WatchdogFileFromNamespaceName(baseDir, namespace, domain)
+
+	return diskutils.RemoveFile(file)
+}
+
+func WatchdogFileUpdate(watchdogFile string) error {
+	f, err := os.Create(watchdogFile)
+	if err != nil {
+		return err
+	}
+	f.Close()
+
+	return nil
+}
+
+func WatchdogFileExists(baseDir string, vm *v1.VirtualMachine) (bool, error) {
+	namespace := precond.MustNotBeEmpty(vm.GetObjectMeta().GetNamespace())
+	domain := precond.MustNotBeEmpty(vm.GetObjectMeta().GetName())
+
+	filePath := WatchdogFileFromNamespaceName(baseDir, namespace, domain)
+	exists, err := diskutils.FileExists(filePath)
+	if err != nil {
+		return false, err
```

Comment:
It might be helpful to add contextual messages to these errors while attempting to manipulate the watchdog file.

Reply:
fixed